File: C++/dfa.cpp
#include <bits/stdc++.h>
#include "dfa.hpp"

using namespace std;

int present_in(vector<string> alphabet, string s) {
  if(find(alphabet.begin(), alphabet.end(), s) != alphabet.end()) return 1;
  else return 0;
}

State::State(bool s_start, bool s_final, string s_name)
  : is_start(s_start), is_final(s_final), name(s_name) {}

DFA::DFA(vector<string> alphabet)
  : sigma(alphabet), rejected(0), curr_state(nullptr) {}


void DFA::add_state(bool is_start, bool is_final, string name) {
  if(states.find(name) == states.end()) {
    State *state = new State(is_start, is_final, name);
    states[name] = state;
  } else {
    cout << "State already exists!";
  }

}

void DFA::add_transition(string curr_state, string input_symbol, string next_state) {
  if(!present_in(sigma, input_symbol)) {
    cout << "Input symbol not part of the alphabet";
    return;
  }

  if(states.find(curr_state) == states.end()) {
    cout << "Current state is not part of the existing set of states";
    return;
  }

  if(states.find(next_state) == states.end()) {
    cout << "Next state is not part of the existing set of states";
    return;
  }

  if(transitions.find(input_symbol) == transitions.end()) {
    transitions[input_symbol] = vector<tuple<string, string, string>> ();
  }

  transitions[input_symbol].push_back(make_tuple(curr_state, input_symbol, next_state));
}

State* DFA::get_start_state() {
  for(auto i : states) {
    if((i.second)->is_start == true) return i.second;
  }
}

vector<string> DFA::get_final_states() {
  vector<string> final_s;

  for(auto i : states) {
    if((i.second)->is_final == true) final_s.push_back(i.second->name);
  }
  return final_s;
}

void DFA::apply_transition(vector<string> partial_transition) {
  string symbol = partial_transition[1];
  string current = partial_transition[0];
  
  vector<tuple<string, string, string>> total;
  total = transitions[symbol];
  vector<tuple<string, string, string>> applicable;

  for(int i = 0;i < total.size(); i++) {
    if(get<0>(total[i]) == current) applicable.push_back(total[i]);
  }

  if(applicable.size()) {
    curr_state = states[get<2>(applicable[0])];
  } else {
    rejected = 1;
    curr_state = states[current];
  }
}

void DFA::evaluate_string(string str) {
  for(int i = 0; i < str.length(); i++) {
    string symb(1, str[i]);
    if(!present_in(sigma, symb)) {
      cout << str[i] << " is not part of the alphabet\n";
      return;
    }
  }

  rejected = 0;
  State* start = get_start_state();
  vector<string> final_states;
  final_states = get_final_states();

  curr_state = start;

  for(int i = 0; i < str.length(); i++) {
    if(rejected == 0) {
      string symb(1, str[i]);
      vector<string> partial = {curr_state->name, symb};
      apply_transition(partial);
    } else {
      cout << "Rejected\n";
      return;
    }
  }

  if(find(final_states.begin(), final_states.end(), curr_state->name) != final_states.end())
    cout << "Accepted!\n";

  else
    cout << "Rejected\n";
}

File: Nim/dfa.nim
import tables

type State = ref object
    is_start : bool
    is_final : bool
    name : string

type DFA = object
    sigma : seq[string]
    transitions : OrderedTableRef[string, seq[(string, string, string)]]
    states : OrderedTableRef[string, State]
    rejected : bool
    curr_state : State


proc new_DFA(alphabet : seq[string]): ref DFA =
    var dfa : ref DFA = new(DFA)
    dfa.sigma = alphabet
    dfa.rejected = false
    dfa.curr_state = nil
    dfa.states = newOrderedTable[string, State]()
    dfa.transitions = newOrderedTable[string, seq[(string, string, string)]]()
    return dfa

proc add_DFA_state(dfa : ref DFA, is_start : bool, is_final : bool, name : string): void =
    if not dfa.states.contains(name):
        let state : State = State(is_start : is_start, is_final : is_final, name : name)
        dfa.states[name] = state
    else:
        echo "State already exists"

proc add_DFA_transition(dfa : ref DFA, curr_state : string, input_symbol : string, next_state : string): void =
    if not (input_symbol in dfa.sigma):
        echo "Input symbol not part of the alphabet"
        return
    if not dfa.states.contains(curr_state):
        echo "Current state is not part of the existing set of states"
        return
    if not dfa.states.contains(next_state):
        echo "Next state is not part of the existing set of states"
        return
    if not dfa.transitions.contains(input_symbol):
        dfa.transitions[input_symbol] = newSeq[(string, string, string)]()
    dfa.transitions[input_symbol].add( (curr_state, input_symbol, next_state) )

proc get_DFA_start_state(dfa : ref DFA): State =
    for key, value in dfa.states.pairs:
        if value.is_start == true:
            return value

proc get_DFA_final_states(dfa : ref DFA): seq[string] =
    var final_s : seq[string] = newSeq[string]()
    for key, value in dfa.states.pairs:
        if value.is_final == true:
            final_s.add(value.name)
    return final_s


proc apply_DFA_transition(dfa : ref DFA, partial_transition : seq[string]): void =
    let
        symbol : string = partial_transition[1]
        current : string = partial_transition[0]
        total : seq[(string, string, string)] = dfa.transitions[symbol]

    var applicable : seq[(string, string, string)]

    for i in countup(0, total.len()-1):
        if total[i][0] == current:
            applicable.add(total[i])

    if applicable.len() != 0:
        dfa.curr_state = dfa.states[applicable[0][2]]
    else:
        dfa.rejected = true
        dfa.curr_state = dfa.states[current]


proc evaluate_DFA_string(dfa : ref DFA, str : string): void =
    for i in countup(0, str.len()-1):
        var symb: string
        symb.add(str[i])
        if not (symb in dfa.sigma):
            echo symb & " is not part of the alphabet"
            return

    dfa.rejected = false
    let start : State = dfa.get_DFA_start_state()
    let final : seq[string] = dfa.get_DFA_final_states()

    dfa.curr_state = start

    for i in countup(0, str.len()-1):
        if not dfa.rejected:
            var symb: string
            symb.add(str[i])
            let partial : seq[string] = @[dfa.curr_state.name, symb]
            dfa.apply_DFA_transition(partial)
        else:
            echo "Rejected"
            return

    if final.contains(dfa.curr_state.name):
        echo "Accepted!"
    else:
        echo "Rejected"

let alphabet : seq[string] = @["a", "b"]
let dfa : ref DFA = new_DFA(alphabet)

dfa.add_DFA_state(true, true, "q0")
dfa.add_DFA_state(false, true, "q1")
dfa.add_DFA_state(false, true, "q2")
dfa.add_DFA_state(false, true, "q3")
dfa.add_DFA_state(false, false, "q4")

dfa.add_DFA_transition("q0", "a", "q1")
dfa.add_DFA_transition("q0", "b", "q0")
dfa.add_DFA_transition("q1", "a", "q2")
dfa.add_DFA_transition("q1", "b", "q1")
dfa.add_DFA_transition("q2", "a", "q3")
dfa.add_DFA_transition("q2", "b", "q2")
dfa.add_DFA_transition("q3", "a", "q4")
dfa.add_DFA_transition("q3", "b", "q3")
dfa.add_DFA_transition("q4", "a", "q4")
dfa.add_DFA_transition("q4", "b", "q4")

dfa.evaluate_DFA_string("aaab")
dfa.evaluate_DFA_string("aaaab")
dfa.evaluate_DFA_string("")
dfa.evaluate_DFA_string("bbbb")